[Kernel]  Expand MoE weight loading + Add Fused Marlin MoE Kernel

dsikka opened the pull request:
# Summary
- Expands weight loading to support grouped and per channel weight quantization. Cleans-up fp8 MoE to use the updated weight loading
- Adds Marlin Fused MoE Kernel for `w4a16` by @ElizaWszola 
- Adds `CompressedTensorsMoEMethod` to support MoE `w4a16` models from [llm-compressor](https://github.com/vllm-project/llm-compressor) and [compressed-tensors](https://github.com/neuralmagic/compressed-tensors)
- Tested using 2 and 4 TP with Mixtral

TODO:
- Test with deepseek v2

Next Steps:
- The `CompressedTensorsMoEMethod` is not leveraging the scheme structure in-place for compressed-tensors in order to keep the scope of this PR focused on the kernel + updated weight loading. Will be updated in a follow-up to use the scheme structure 

co-authored by @ElizaWszola, from Neural Magic

dsikka committed:
clean-up model weight loading; support channel wise weight scales

dsikka committed:
update fp8 MoE to use updated weight loading

robertgshaw2-neuralmagic commented on vllm/model_executor/layers/fused_moe/layer.py:
```diff
@@ -199,21 +205,41 @@ def __init__(
-            self._load_fp8_scale(param.data, loaded_weight, weight_name,
-                                 shard_id, expert_id)
-            return
+    def _load_per_tensor_weight_scale(self, shard_id: str,
```
could these be utilities that can be shared with autogptq and autoawq?

dsikka replied:
Yep

robertgshaw2-neuralmagic commented on vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:
```diff
@@ -400,3 +404,64 @@ def validate_kv_cache_scheme(kv_cache_scheme: Optional[Dict[str, Any]]):
+                       hidden_size: int, intermediate_size: int,
+                       params_dtype: torch.dtype, **extra_weight_attrs):
+
+        w13_weight = torch.nn.Parameter(torch.empty(num_experts,
```
I would add some comments about: 

- w1 being gate_proj
- w3 being up_proj
- w2 being down_proj

robertgshaw2-neuralmagic commented on vllm/model_executor/layers/quantization/fp8.py:
```diff
@@ -306,11 +306,11 @@ def create_weights(self, layer: Module, num_experts: int, hidden_size: int,
         if self.quant_config.is_checkpoint_fp8_serialized:
             set_weight_attrs(w13_weight_scale, {
-                "is_fp8_scale": True,
+                "quant_method": "tensor",
```
I think we should avoid this is possible

Would the VLLMParameters be able to support this?

dsikka replied:
Yes, I just didnt want to bring it in yet so this was a quick workaround

dsikka committed:
add group quant weight load support

dsikka committed:
moe kernel (#403)

FILL IN THE PR DESCRIPTION HERE

FIX #xxxx (*link existing issues this PR will resolve*)

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE
DESCRIPTION ABOVE**

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so
we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull
request, please ensure the PR meets the following criteria. This helps
vLLM maintain the code quality and improve the efficiency of the review
process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed
appropriately to indicate the type of change. Please use one of the
following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
<li><code>[CI/Build]</code> for build or continuous integration
improvements.</li>
<li><code>[Doc]</code> for documentation fixes and improvements.</li>
<li><code>[Model]</code> for adding a new model or improving an existing
model. Model name should appear in the title.</li>
<li><code>[Frontend]</code> For changes on the vLLM frontend (e.g.,
OpenAI API server, <code>LLM</code> class, etc.) </li>
<li><code>[Kernel]</code> for changes affecting CUDA kernels or other
compute kernels.</li>
<li><code>[Core]</code> for changes in the core vLLM logic (e.g.,
<code>LLMEngine</code>, <code>AsyncLLMEngine</code>,
<code>Scheduler</code>, etc.)</li>
<li><code>[Hardware][Vendor]</code> for hardware-specific changes.
Vendor name should appear in the prefix (e.g.,
<code>[Hardware][AMD]</code>).</li>
<li><code>[Misc]</code> for PRs that do not fit the above categories.
Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please
include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
<li>We adhere to <a
href="https://google.github.io/styleguide/pyguide.html">Google Python
style guide</a> and <a
href="https://google.github.io/styleguide/cppguide.html">Google C++
style guide</a>.</li>
<li>Pass all linter checks. Please use <a
href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a>
to format your code.</li>
<li>The code need to be well-documented to ensure future contributors
can easily understand the code.</li>
<li>Include sufficient tests to ensure the project to stay correct and
robust. This includes both unit tests and integration tests.</li>
<li>Please add documentation to <code>docs/source/</code> if the PR
modifies the user-facing behaviors of vLLM. It helps vLLM user
understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major
architectural changes (>500 LOC excluding kernel/data/config/test), we
would expect a GitHub issue (RFC) discussing the technical design and
justification. Otherwise, we will tag it with <code>rfc-required</code>
and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing
machine</i>. We would like to make the review process transparent and
efficient and make sure no contributor feel confused or frustrated.
However, the vLLM team is small, so we need to prioritize some PRs over
others. Here is what you can expect from the review process: </p>

<ul>
<li> After the PR is submitted, the PR will be assigned to a reviewer.
Every reviewer will pick up the PRs based on their expertise and
availability.</li>
<li> After the PR is assigned, the reviewer will provide status update
every 2-3 days. If the PR is not reviewed within 7 days, please feel
free to ping the reviewer or the vLLM team.</li>
<li> After the review, the reviewer will put an <code>
action-required</code> label on the PR if there are changes required.
The contributor should address the comments and ping the reviewer to
re-review the PR.</li>
<li> Please respond to all comments within a reasonable time frame. If a
comment isn't clear or you disagree with a suggestion, feel free to ask
for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and
for your interest in contributing to vLLM. Your contributions make vLLM
a great tool for everyone! </p>

</details>

dsikka committed:
clean-up

dsikka committed:
more clean-up

dsikka committed:
update condition for arch

dsikka committed:
update deepseek and qwen to skip weight_shape loading

dsikka committed:
comments

mgoin commented on vllm/model_executor/models/deepseek_v2.py:
```diff
@@ -553,6 +553,8 @@ def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
 
         params_dict = dict(self.named_parameters())
         for name, loaded_weight in weights:
+            if "weight_shape" in name:
```
Could you make this more specific? At least should be able to upgrade to `.weight_shape`

Also what is this added from?

dsikka replied:
This was added because compressed-tensors gives us weight_shapes which we never use. In all the other cases, we just load the parameter and then dont use it during the forward pass. Added this as a quick workaround but will add the parameters to be consistent.

mgoin reviewed (commented):
Nice work, I'm liking the start of refactoring in fused_moe/layer.py

dsikka committed:
format

dsikka commented:
/ready

dsikka committed:
fix issue on upstream-main

dsikka commented on vllm/_custom_ops.py:
```diff
@@ -306,6 +306,19 @@ def awq_marlin_repack(b_q_weight: torch.Tensor, size_k: int, size_n: int,
+                           size_k: int, size_n: int,
+                           num_bits: int) -> torch.Tensor:
+    num_experts = b_q_weight.shape[0]
+    output = torch.empty((num_experts, size_k // 16, size_n * 2),
```
@ElizaWszola

ElizaWszola replied:
yes, I think it's a reasonable assertion

ElizaWszola replied:
done!

dsikka committed:
fix imports for tests

dsikka commented on vllm/model_executor/layers/fused_moe/layer.py:
```diff
@@ -199,55 +206,172 @@ def __init__(
+            idx = 0 if shard_id == "w1" else 1
+            param_data[expert_id][idx] = loaded_weight
+        # If we are in the row parallel case (down_proj)
+        else:
```
there's one exception at the very beginning of the weight_loader function. so it would be raised before we get to this point, if shard_id is not one of w1,w2,w3

dsikka commented on vllm/model_executor/layers/fused_moe/layer.py:
```diff
@@ -199,55 +206,172 @@ def __init__(
+                           shard_dim=shard_dim,
+                           loaded_weight=loaded_weight,
+                           expert_data=expert_data,
+                           tp_rank=tp_rank)
```
see above

dsikka commented on vllm/model_executor/layers/fused_moe/layer.py:
```diff
@@ -199,55 +206,172 @@ def __init__(
+        ]
+        # Fetch the dim to shard the parameter/loaded weight
+        # based on the shard id
+        SHARD_ID_TO_SHARDED_DIM = {"w1": 0, "w2": 1, "w3": 0}
```
Yes. w1 and w3 are fused together (2 logical matrices, one actual matrix) so it it's the same dimension. I will make a note that the dimension to shard is whatever dimension the `intermediate_size` is used.

dsikka commented on vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:
```diff
@@ -400,3 +408,256 @@ def validate_kv_cache_scheme(kv_cache_scheme: Optional[Dict[str, Any]]):
                 f"However found symmetric: {is_symmetric}")
+
+
+class GPTQMarlinState(Enum):
+    REPACK = enum.auto()
+    READY = enum.auto()
+
+
+class CompressedTensorsMoEMethod(FusedMoEMethodBase):
```
yep, I agree

dsikka commented on vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:
```diff
@@ -400,3 +408,256 @@ def validate_kv_cache_scheme(kv_cache_scheme: Optional[Dict[str, Any]]):
+              num_expert_group: Optional[int] = None,
+              topk_group: Optional[int] = None) -> torch.Tensor:
+
+        # hook-up fused moe kernel
+        if layer.marlin_state == GPTQMarlinState.REPACK:
+            layer.marlin_state = GPTQMarlinState.READY
```
@ElizaWszola

ElizaWszola replied:
done!

dsikka commented on vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:
```diff
@@ -400,3 +408,256 @@ def validate_kv_cache_scheme(kv_cache_scheme: Optional[Dict[str, Any]]):
+        self.num_bits = config.num_bits
+        self.packed_factor = 32 // config.num_bits
+        self.strategy = config.strategy.value
+        self.group_size = config.group_size
```
we can check symmetric. will add that in

dsikka committed:
PR comments + fix import issue

dsikka committed:
format

ElizaWszola committed:
Michael's feedback

mgoin commented on vllm/model_executor/layers/fused_moe/fused_moe.py:
```diff
@@ -323,21 +323,16 @@ def get_moe_configs(E: int, N: int,
         'GROUP_SIZE_M': 8
     }
-    if M <= E:
+    if M <= E or (is_marlin and M <= 32):
```
Is `M <= 32` because we dequant fully outside of that case?

dsikka replied:
@ElizaWszola

ElizaWszola replied:
It's a heuristic aimed at performance: marlin kernels are faster with this blocking for `M <= 32`

mgoin commented on vllm/model_executor/layers/fused_moe/fused_moe.py:
```diff
@@ -441,6 +437,104 @@ def grouped_topk(hidden_states: torch.Tensor,
+                                        w1.shape,
+                                        w2.shape,
+                                        topk_ids.shape[1],
+                                        "float8" if use_fp8 else None,
```
Shouldn't we assert that `use_fp8 is None`?

ElizaWszola replied:
done

dsikka committed:
testing

ElizaWszola committed:
Assert no fp8 in fused moe

dsikka committed:
move to weight_loading test to deal with CI cuda memory issues

fengyang95 commented:
Have you tested deepseek-v2?

fengyang95 commented:
I tried to run deepseek-v2-fp8(https://huggingface.co/neuralmagic/DeepSeek-Coder-V2-Instruct-FP8) with this PR and 8 L40 GPUs, and encountered the following error:
model: 
VllmWorkerProcess pid=6252) INFO 08-22 22:07:58 model_runner.py:890] Loading model weights took 28.2205 GB
(VllmWorkerProcess pid=6256) INFO 08-22 22:07:58 model_runner.py:890] Loading model weights took 28.2205 GB
(VllmWorkerProcess pid=6255) INFO 08-22 22:07:58 model_runner.py:890] Loading model weights took 28.2205 GB
(VllmWorkerProcess pid=6254) INFO 08-22 22:07:58 model_runner.py:890] Loading model weights took 28.2205 GB
(VllmWorkerProcess pid=6257) INFO 08-22 22:07:58 model_runner.py:890] Loading model weights took 28.2205 GB
INFO 08-22 22:07:58 model_runner.py:890] Loading model weights took 28.2205 GB
(VllmWorkerProcess pid=6251) INFO 08-22 22:07:58 model_runner.py:890] Loading model weights took 28.2205 GB
(VllmWorkerProcess pid=6253) INFO 08-22 22:07:58 model_runner.py:890] Loading model weights took 28.2205 GB
ERROR 08-22 22:07:59 multiproc_worker_utils.py:120] Worker VllmWorkerProcess pid 6252 died, exit code: -15
INFO 08-22 22:07:59 multiproc_worker_utils.py:123] Killing local vLLM worker processes
Process SpawnProcess-1:
Traceback (most recent call last):
  File "/usr/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/usr/lib/python3.9/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.9/dist-packages/vllm/entrypoints/openai/rpc/server.py", line 230, in run_rpc_server
    server = AsyncEngineRPCServer(async_engine_args, usage_context, rpc_path)
  File "/usr/local/lib/python3.9/dist-packages/vllm/entrypoints/openai/rpc/server.py", line 31, in __init__
    self.engine = AsyncLLMEngine.from_engine_args(
  File "/usr/local/lib/python3.9/dist-packages/vllm/engine/async_llm_engine.py", line 740, in from_engine_args
    engine = cls(
  File "/usr/local/lib/python3.9/dist-packages/vllm/engine/async_llm_engine.py", line 636, in __init__
    self.engine = self._init_engine(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/vllm/engine/async_llm_engine.py", line 840, in _init_engine
    return engine_class(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/vllm/engine/async_llm_engine.py", line 272, in __init__
    super().__init__(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/vllm/engine/llm_engine.py", line 284, in __init__
    self._initialize_kv_caches()
  File "/usr/local/lib/python3.9/dist-packages/vllm/engine/llm_engine.py", line 390, in _initialize_kv_caches
    self.model_executor.determine_num_available_blocks())
  File "/usr/local/lib/python3.9/dist-packages/vllm/executor/distributed_gpu_executor.py", line 38, in determine_num_available_blocks
    num_blocks = self._run_workers("determine_num_available_blocks", )
  File "/usr/local/lib/python3.9/dist-packages/vllm/executor/multiproc_gpu_executor.py", line 192, in _run_workers
    driver_worker_output = driver_worker_method(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/vllm/worker/worker.py", line 222, in determine_num_available_blocks
    self.model_runner.profile_run()
  File "/usr/local/lib/python3.9/dist-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/vllm/worker/model_runner.py", line 1097, in profile_run
    self.execute_model(model_input, kv_caches, intermediate_tensors)
  File "/usr/local/lib/python3.9/dist-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/vllm/worker/model_runner.py", line 1415, in execute_model
    hidden_or_intermediate_states = model_executable(
  File "/usr/local/lib/python3.9/dist-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/vllm/model_executor/models/deepseek_v2.py", line 504, in forward
    hidden_states = self.model(input_ids, positions, kv_caches,
  File "/usr/local/lib/python3.9/dist-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/vllm/model_executor/models/deepseek_v2.py", line 461, in forward
    hidden_states, residual = layer(positions, hidden_states,
  File "/usr/local/lib/python3.9/dist-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/vllm/model_executor/models/deepseek_v2.py", line 401, in forward
    hidden_states = self.mlp(hidden_states)
  File "/usr/local/lib/python3.9/dist-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/vllm/model_executor/models/deepseek_v2.py", line 148, in forward
    final_hidden_states = self.experts(
  File "/usr/local/lib/python3.9/dist-packages/torch/nn/modules/module.py", line 1553, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/torch/nn/modules/module.py", line 1562, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/vllm/model_executor/layers/fused_moe/layer.py", line 287, in forward
    final_hidden_states = self.quant_method.apply(
  File "/usr/local/lib/python3.9/dist-packages/vllm/model_executor/layers/quantization/fp8.py", line 499, in apply
    return fused_experts(x,
  File "/usr/local/lib/python3.9/dist-packages/vllm/model_executor/layers/fused_moe/fused_moe.py", line 544, in fused_experts
    moe_align_block_size(curr_topk_ids, config['BLOCK_SIZE_M'], E))
  File "/usr/local/lib/python3.9/dist-packages/vllm/model_executor/layers/fused_moe/fused_moe.py", line 228, in moe_align_block_size
    ops.moe_align_block_size(topk_ids, num_experts, block_size, sorted_ids,
  File "/usr/local/lib/python3.9/dist-packages/vllm/_custom_ops.py", line 28, in wrapper
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/vllm/_custom_ops.py", line 485, in moe_align_block_size
    torch.ops._C.moe_align_block_size(topk_ids, num_experts, block_size,
  File "/usr/local/lib/python3.9/dist-packages/torch/_ops.py", line 1061, in __call__
    return self_._op(*args, **(kwargs or {}))
RuntimeError: CUDA error: invalid argument
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.

mgoin commented:
@fengyang95 it looks like your issue is due to incorrectly compiled or linked functions, possibly from older builds. Could you build vllm in a fresh environment?

fengyang95 commented:
> ooks like your issue is due to incorrectly compiled or linked functions, possibly from older builds. Could you build vllm in a fresh environment?

I tried recompiling in a new environment but still encountered the same issue. Has this been tested on ds-v2?

fengyang95 commented:
> > ooks like your issue is due to incorrectly compiled or linked functions, possibly from older builds. Could you build vllm in a fresh environment?
> 
> @mgoin I tried recompiling in a new environment but still encountered the same issue. Has this been tested on ds-v2?

dsikka commented:
@fengyang95 Do you mind sharing a code snippet of what you're running?
The scope of the fused marlin kernel  introduced in this PR is focused on w4a16 models.  It does not support FP8 models at the moment.

fengyang95 commented:
> @fengyang95 Do you mind sharing a code snippet of what you're running? The scope of the fused marlin kernel introduced in this PR is packed w4a16 models. It does not support FP8 models at the moment.
Oh, I was using an FP8 model.
https://huggingface.co/neuralmagic/DeepSeek-Coder-V2-Instruct-FP8

dsikka commented:
@fengyang95  FP8 models will use the triton kernels which should work with the DeepSeek V2 model listed.
Adding FP8 Fused MoE Marlin kernel is not in scope at the moment.

fengyang95 commented:
How should I load a w4a16 version of deepseek-v2 that was compressed using llm-compressor?
